frontend/src/metabase-lib/lib/queries/StructuredQuery.ts

```diff
@@ -307,23 +307,30 @@ export default class StructuredQuery extends AtomicQuery {
         name: "",
         display_name: "",
         db: sourceQuery.database(),
-        fields: sourceQuery.columns().map(
-          column =>
-            new Field({
-              ...column,
-              // TODO FIXME -- Do NOT use field-literal unless you're referring to a native query
-              id: [
-                "field",
-                column.name,
-                {
-                  "base-type": column.base_type,
-                },
-              ],
-              source: "fields",
-              // HACK: need to thread the query through to this fake Field
-              query: this,
-            }),
-        ),
+        fields: sourceQuery.columns().map(column => {
+          const hasMappingField = this.question()
```

Comment:
We are trying to find a `column` (using `base_type` and `name`) matching the `field`.

Reply:
I'm pretty sure `sourceQuery.columns()` relies on the question/card's resultMetadata under the hood, so I don't think doing this is exactly what we want. If you add a custom column to the question this prevents us from properly mapping to it as it has an `id` that is an array... fun, right?

It _seems_ (I am never 100% confident about changing anything having to do with this stuff) that if you replace `sourceQuery.columns().map(column => {...})` with `sourceQuery.dimension().map(dimension => dimension.field())` you can avoid this fake field nonsense entirely--StructuredQuery shouldn't be getting involved with field generation--we should aim for a better separation of concerns here--so `Dimension` should handle the tedious aspects of our `Field` logic, and it _mostly_ does, and where it doesn't it suggests that we need to improve upon our Dimensions logic so that we can rely on it more going forward.

I've found at least a few problems with this approach, unfortunately, as there are myriad exceptions to the `Dimension` logic that exists already, despite its complexity, so any time we use `Dimension` instances in new ways we find problems. It seems like the `field` method on `ExpressionDimension` has some trouble parsing some custom expressions, so I'm going to suggest an additional change to account for that.

This switch statement found here https://github.com/metabase/metabase/blob/00b6d1c8eb11b7e1d3017cf2db0acd2de99c9e38/frontend/src/metabase-lib/lib/Dimension.ts#L1263 has cases like `MONOTYPE.String` that map to strings with capital letters ("String" for `MONOTYPE.String`) but it appears that the `type` value it is evaluating against can be an all lowercase `"string"` if it ends up getting set as a `type` property from https://github.com/metabase/metabase/blob/00b6d1c8eb11b7e1d3017cf2db0acd2de99c9e38/frontend/src/metabase/lib/expressions/config.js#L70. I think to fix this we should probably do something like:

```
      switch (type.toLowerCase()) {
        case MONOTYPE.String.toLowerCase():
          base_type = "type/Text";
          break;

        case MONOTYPE.Boolean.toLowerCase():
          base_type = "type/Boolean";
          break;

        default:
          break;
      }
```

Probably worth pinging @ariya about this logic in case there's something that makes more sense.

In addition to this problem, there is a second problem relating to the `ExpressionDimension.prototype.field` method running a second time but with an expression that has been turned into an array. The existing inference logic breaks on the array. Instead of going through the parsing logic again, it seems like we can rely on the fact that the custom column exists in the associated question's `result_metadata` array, so we can add something like this above all the inference logic in `ExpressionDimension.prototype.field` to grab that `result_metadata` field:

```
    if (query?.question()?.getResultMetadata?.()) {
      const resultMetadata = query.question().getResultMetadata();
      const field = _.findWhere(resultMetadata, {
        name: this._expressionName,
      });

      if (field) {
        return new Field({
          ...field,
          metadata: this._metadata,
          query: this._query,
        });
      }
    }
```

These changes need a LOT of testing because there's a lot that can go wrong. I'm sure we might need to tweak more logic relating to this stuff.